src/indentblock.js

```diff
@@ -0,0 +1,289 @@
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module indent-block/indentblock
+ */
+
+import Plugin from '@ckeditor/ckeditor5-core/src/plugin';
+
+import IndentBlockCommand from './indentblockcommand';
+import IndentUsingOffset from './indentcommandbehavior/indentusingoffset';
+import IndentUsingClasses from './indentcommandbehavior/indentusingclasses';
+
+/**
+ * The block indentation feature.
+ *
+ * It registers the `'indentBlock'` and `'outdentBlock'` commands.
+ *
+ * If the plugin {@link module:core/indent/indent~Indent} is defined it also attaches the `'indentBlock'` and `'outdentBlock'` commands to
+ * `'indent'` and `'outdent'` commands.
+ *
+ * @extends module:core/plugin~Plugin
+ */
+export default class IndentBlock extends Plugin {
+	/**
+	 * @inheritDoc
+	 */
+	constructor( editor ) {
+		super( editor );
+
+		editor.config.define( 'indentBlock', {
+			offset: 1,
+			unit: 'em'
+		} );
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	static get pluginName() {
+		return 'IndentBlock';
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	init() {
+		const editor = this.editor;
+		const schema = editor.model.schema;
+		const conversion = editor.conversion;
+		const configuration = editor.config.get( 'indentBlock' );
+
+		// TODO: supported in CKE4:
+		//  - p,
+		//  - h1, h2, h3, h4, h5, h6,
+		//  - ul, ol,
+		//  - div,
+		//  - dl,
+		//  - pre,
+		//  - table
+		const knownElements = [ 'paragraph', 'heading1', 'heading2', 'heading3', 'heading4', 'heading5', 'heading6' ];
+
+		knownElements.forEach( elementName => {
+			if ( schema.isRegistered( elementName ) ) {
+				schema.extend( elementName, { allowAttributes: 'blockIndent' } );
+			}
+		} );
+
+		const useOffsetConfig = !configuration.classes || !configuration.classes.length;
+
+		const indentConfig = Object.assign( { direction: 'forward' }, configuration );
+		const outdentConfig = Object.assign( { direction: 'backward' }, configuration );
+
+		if ( useOffsetConfig ) {
+			this._setupConversionUsingOffset( conversion );
+			editor.commands.add( 'indentBlock', new IndentBlockCommand( editor, new IndentUsingOffset( indentConfig ) ) );
+			editor.commands.add( 'outdentBlock', new IndentBlockCommand( editor, new IndentUsingOffset( outdentConfig ) ) );
+		} else {
+			this._setupConversionUsingClasses( configuration.classes );
+			editor.commands.add( 'indentBlock', new IndentBlockCommand( editor, new IndentUsingClasses( indentConfig ) ) );
+			editor.commands.add( 'outdentBlock', new IndentBlockCommand( editor, new IndentUsingClasses( outdentConfig ) ) );
+		}
+
+		const getCommandExecuter = commandName => {
+			return ( data, cancel ) => {
+				const command = this.editor.commands.get( commandName );
+
+				if ( command.isEnabled ) {
+					this.editor.execute( commandName );
+					cancel();
+				}
+			};
+		};
+
+		editor.keystrokes.set( 'Tab', getCommandExecuter( 'indentBlock' ) );
```

Comment:
<kbd>Tab</kbd> should only work in lists. For a11y reasons, by default, it should leave the editor. So this code should be removed.